components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/engine/CamelSubscription.java

```diff
@@ -201,46 +201,52 @@ public void cancel() {
     }
 
-    protected void discardBuffer(List<StreamPayload<Exchange>> remaining) {
-        for (StreamPayload<Exchange> data : remaining) {
-            data.getCallback().processed(data.getItem(), new IllegalStateException("Cannot process the exchange " + data.getItem() + ": subscription cancelled"));
+    protected void discardBuffer(List<Exchange> remaining) {
+        for (Exchange data : remaining) {
+            DispatchCallback<Exchange> callback = ReactiveStreamsHelper.getCallback(data);
+            if (callback != null) {
+                callback.processed(data, new IllegalStateException("Cannot process the exchange " + data + ": subscription cancelled"));
+            }
         }
     }
 
-    public void publish(StreamPayload<Exchange> message) {
-        Map<StreamPayload<Exchange>, String> discardedMessages = null;
+    public void publish(Exchange message) {
+        Map<Exchange, String> discardedMessages = null;
         try {
             mutex.lock();
             if (!this.terminating && !this.terminated) {
-                Collection<StreamPayload<Exchange>> discarded = this.backpressureStrategy.update(buffer, message);
+                Collection<Exchange> discarded = this.backpressureStrategy.update(buffer, message);
                 if (discarded.iterator().hasNext()) {
                     discardedMessages = new HashMap<>();
-                    for (StreamPayload<Exchange> ex : discarded) {
-                        discardedMessages.put(ex, "Exchange " + ex.getItem() + " discarded by backpressure strategy " + this.backpressureStrategy);
+                    for (Exchange ex : discarded) {
+                        discardedMessages.put(ex, "Exchange " + ex + " discarded by backpressure strategy " + this.backpressureStrategy);
                     }
                 }
             } else {
                 // acknowledge
-                discardedMessages = Collections.singletonMap(message, "Exchange " + message.getItem() + " discarded: subscription closed");
+                discardedMessages = Collections.singletonMap(message, "Exchange " + message + " discarded: subscription closed");
             }
         } finally {
             mutex.unlock();
         }
 
         // discarding outside of mutex scope
         if (discardedMessages != null) {
-            for (Map.Entry<StreamPayload<Exchange>, String> discarded : discardedMessages.entrySet()) {
-                StreamPayload<Exchange> m = discarded.getKey();
-                Exchange exchange = m.getItem();
-                ReactiveStreamsDiscardedException e = new ReactiveStreamsDiscardedException("Discarded by backpressure strategy", exchange, name);
-                m.getCallback().processed(exchange, e);
+            for (Map.Entry<Exchange, String> discarded : discardedMessages.entrySet()) {
+                Exchange exchange = discarded.getKey();
+                DispatchCallback<Exchange> callback = ReactiveStreamsHelper.getCallback(exchange);
```

Comment:
It is an option but it would create a new object per event without a real benefit.